### PR TITLE
Add podman volume mount support

### PR DESCRIPTION
**Pull request by rhatdan**
Fixes: https://github.com/containers/podman/issues/12768

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

File: test/system/160-volumes.bats [hunk 1]
```diff
@@ -387,4 +387,27 @@ NeedsChown    | true
...
+    is "$output" "$myvolume" "output from volume create"
+
+    if ! is_rootless ; then
+        # image mount is hard to test as a rootless user
+        # and does not work remotely
+        run_podman volume mount ${myvolume}
+        mnt=${output}
+	touch $mnt/$myfile
+        run_podman run -v ${myvolume}:/vol1:z $IMAGE ls /vol1
+	is "$output" "$myfile" "$myfile should exists within the containers volume"
```
Comment (edsantiago) [inline test/system/160-volumes.bats:406]: inconsistent indentation; mix of tabs and spaces. Also, `s/exists/exist/`

File: test/system/160-volumes.bats [hunk 2]
```diff
@@ -387,4 +387,27 @@ NeedsChown    | true
     run_podman volume rm $myvolume
 }
 
+@test "podman volume mount" {
+    skip_if_remote "podman --remote volume mount not supported"
```
Comment (edsantiago) [inline test/system/160-volumes.bats:391]: Test doesn't really do anything when rootless. Why not just `skip_if_rootless`?

File: test/system/160-volumes.bats [hunk 3]
```diff
@@ -387,4 +387,27 @@ NeedsChown    | true
...
+    # Create a named volume
+    run_podman volume create $myvolume
+    is "$output" "$myvolume" "output from volume create"
+
+    if ! is_rootless ; then
+        # image mount is hard to test as a rootless user
+        # and does not work remotely
+        run_podman volume mount ${myvolume}
+        mnt=${output}
+	touch $mnt/$myfile
```
Comment (edsantiago) [inline test/system/160-volumes.bats:404]: Why not `echo $my-random-string-defined-above >$mnt/$myfile` and then `podman run --rm -v ... cat /vol1/$myfile` ?

File: test/system/160-volumes.bats [hunk 4]
```diff
@@ -387,4 +387,27 @@ NeedsChown    | true
...
+        run_podman run -v ${myvolume}:/vol1:z $IMAGE ls /vol1
```
Comment (edsantiago) [inline test/system/160-volumes.bats:405]: Whether you accept my `echo/cat` suggestion or not, please run with `--rm` to avoid littering

**Review by edsantiago — APPROVED** [review]
Nice new feature. Some suggestions for the tests.

**Comment by edsantiago** [comment]
oh, and linter is unhappy
```
pkg/domain/infra/abi/volumes.go:182:1: receiver name ir should be consistent with previous receiver name ic for ContainerEngine (golint)
func (ir *ContainerEngine) VolumeMount(ctx context.Context, nameOrIDs []string) ([]*entities.ContainerMountReport, error) {
^
pkg/domain/infra/abi/volumes.go:198:1: receiver name ir should be consistent with previous receiver name ic for ContainerEngine (golint)
func (ir *ContainerEngine) VolumeUnmount(ctx context.Context, nameOrIDs []string) ([]*entities.ContainerUnmountReport, error) {
^
pkg/domain/infra/abi/auto-update.go:10:28: ST1016: methods on the same type should have the same receiver name (seen 2x "ir", 97x "ic") (stylecheck)
func (ic *ContainerEngine) AutoUpdate(ctx context.Context, options entities.AutoUpdateOptions) ([]*entities.AutoUpdateReport, []error) {
                           ^
make: *** [Makefile:251: golangci-lint] Error 1
```

File: test/system/160-volumes.bats (hunk 4 above)
Comment (rhatdan) [reply test/system/160-volumes.bats]: Fixed

File: test/system/160-volumes.bats (hunk 3 above)
Comment (rhatdan) [reply test/system/160-volumes.bats]: Fixed

File: test/system/160-volumes.bats (hunk 1 above)
Comment (rhatdan) [reply test/system/160-volumes.bats]: Added a small test for rootless.

File: docs/source/markdown/podman-volume-mount.1.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+## SYNOPSIS
+**podman volume mount** [*volume* ...]
+
+## DESCRIPTION
+Mounts the specified volumes' file system in a location which can be
+accessed from the host, and returns its location.
+
+Rootless mode only supports mounting file volumes, unless you enter the user namespace
+via the `podman unshare` command. All other volume type will fail to mount.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-volume-mount.1.md:14]:
```suggestion
via the `podman unshare` command. All other volume types will fail to mount.
```

File: docs/source/markdown/podman-volume-unmount.1.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+## SYNOPSIS
+**podman volume unmount** *volume* [...]
+
+**podman volume umount** *volume* [...]
+
+## DESCRIPTION
+Unmounts the specified volume, if there are no other containers
+using it.
+
+Volume storage increments a mount counter each time an volume is mounted.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-volume-unmount.1.md:15]:
```suggestion
Volume storage increments a mount counter each time a volume is mounted.
```

**Comment by mheon** [comment]
If you do a `podman run` using a volume, then do a `podman volume unmount` on it when the container is running - does that break anything? Want to make sure we don't need to add extra checks around that.

**Comment by rhatdan** [comment]
Since it is in a different mount namespace the unmount will not effect the running container.

Tested this out with a tmpfs container.

**Comment by mheon** [comment]
That isn't true? The volume mounts are made before the container is created, and then we bind-mount the mountpoint into the container - so there is no difference between a container requesting an unmount, and a user requesting an unmount, save for the fact that the container is still potentially running. I'm worried that we'll still set the mount counter to 0 if the unmount fails because the volume is in use, which would result in the volume being considered unmounted when it is still mounted.

**Comment by rhatdan** [comment]
What I see is the mount point in the podman unshare aria shows the tmpfs is no longer mounted, but if I exec into the container I see the tmpfs is still mounted and the data I created on it it is still available.  This is because in the containers mount namespace the file system is still mounted.
If I run another container on the volume, I will see different content because the tmpfs will get mounted again.

**Comment by edsantiago** [comment]
@mheon from my testing, `podman volume unmount` is a NOP. Actually, `volume mount` is, too, all it does is display a path to an existing volume. It does not seem to increment refcnt. podman is perfectly happy to `volume rm` while I still "have" the volume "mounted":
```console
$ sudo bin/podman volume mount myvolume
/var/lib/containers/storage/volumes/myvolume/_data
$ sudo touch files in there, etc etc
$ sudo bin/podman volume rm myvolume
myvolume
$ sudo ls /var/lib/containers/storage/volumes/myvolume
ls: cannot access '/var/lib/containers/storage/volumes/myvolume': No such file or directory
```

**Comment by mheon** [comment]
@edsantiago Did you make the volume with options that would require a mount take place - e.g. `podman volume create --opt type=tmpfs --opt device=tmpfs myvolume` or similar (options may not work perfectly, I'm going off memory here)?

**Comment by edsantiago** [comment]
Uh, no, it was just a plain simple `volume create`. I was just surprised that podman pulled the rug out from under me while I thought I had the volume "mounted". Not a big deal though.

**Comment by mheon** [comment]
Ah - yeah, it's a short-circuit operation unless the volume actually needs a mount. Should probably make that clear in the manpages.

It should always be safe to invoke mount and unmount, even on volumes that don't require a mount, to be clear; it's just that some volumes don't require them, and others do. I think the recommended workflow after the patches would always involve using them.

File: libpod/volume.go [hunk 7]
```diff
@@ -255,3 +255,11 @@ func (v *Volume) IsDangling() (bool, error) {
 func (v *Volume) UsesVolumeDriver() bool {
 	return !(v.config.Driver == define.VolumeDriverLocal || v.config.Driver == "")
 }
+
+func (v *Volume) Mount() (string, error) {
+	return v.config.MountPoint, v.mount()
```
[the remark this answers is not in the document]
Comment (mheon) [reply libpod/volume.go:262]: Thinking about this a bit more - `v.config.MountPoint` may not be populated until `v.mount()` runs (I think this is mostly a volume plugins thing), so it's safer to do `err := v.mount(); return v.config.MountPoint, err`

**Comment by rhatdan** [comment]
I think the bind mount happens after the tmpfs mount which causes the tmpfs mount to leak into the mount namespace.  
Bottom line I did the following
```
$ podman unshare
# ./bin/podman volume create --opt device=tmpfs --opt type=tmpfs --opt o=uid=1000,gid=1000 testvol
# ./bin/podman run -d -v testvol:/vol:Z fedora sleep 1000
52fee416c7a0ac40f085654f4564693d4b5ffc181ec48c174e272378f0283fab
# mount | grep _data
tmpfs on /home/dwalsh/.local/share/containers/storage/volumes/testvol/_data type tmpfs (rw,relatime,seclabel,uid=100999,gid=100999,inode64)
# ./bin/podman exec -l touch /vol/dan
# ./bin/podman volume unmount testvol
testvol
# mount | grep _data
# ./bin/podman exec -l ls /vol
dan
```

File: pkg/domain/infra/abi/volumes.go [hunk 8]
```diff
@@ -178,3 +178,35 @@ func (ic *ContainerEngine) VolumeMounted(ctx context.Context, nameOrID string) (
...
+		} else {
+			report.Path, report.Err = vol.Mount()
+		}
+		reports = append(reports, &report)
+	}
+
+	return reports, nil
+}
+
+func (ic *ContainerEngine) VolumeUnmount(ctx context.Context, nameOrIDs []string) ([]*entities.ContainerUnmountReport, error) {
```
Comment (cdoern) [inline pkg/domain/infra/abi/volumes.go:198]: Same as above, type consistency.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL

**Review by giuseppe — APPROVED** [review]
LGTM

**Review by vrothberg — COMMENTED** [review]
/lgtm

**Pull request merged**